pkgs/applications/networking/instant-messengers/telegram/tdesktop/generic.nix

```diff
@@ -0,0 +1,198 @@
+{ mkDerivation
+, lib
+, fetchFromGitHub
+, callPackage
+, pkg-config
+, cmake
+, ninja
+, python3
+, wrapGAppsHook
+, wrapQtAppsHook
+, extra-cmake-modules
+, qtbase
+, qtimageformats
+, gtk3
+, kwayland
+, libdbusmenu
+, lz4
+, xxHash
+, ffmpeg
+, openalSoft
+, minizip
+, libopus
+, alsa-lib
+, libpulseaudio
+, range-v3
+, tl-expected
+, hunspell
+, glibmm
+, webkitgtk
+, jemalloc
+, rnnoise
+, abseil-cpp
+  # Transitive dependencies:
+, util-linuxMinimal
+, pcre
+, libpthreadstubs
+, libXdmcp
+, libselinux
+, libsepol
+, libepoxy
+, at-spi2-core
+, libXtst
+, libthai
+, libdatrie
+, xdg-utils
+, libsysprof-capture
+, libpsl
+, brotli
+, microsoft_gsl
+, rlottie
+, stdenv
+, gcc10Stdenv
+  # Version-specific attributes
+, pname
+, version
+, sha256
+}:
+
+# Main reference:
+# - This package was originally based on the Arch package but all patches are now upstreamed:
+#   https://git.archlinux.org/svntogit/community.git/tree/trunk/PKGBUILD?h=packages/telegram-desktop
+# Other references that could be useful:
+# - https://git.alpinelinux.org/aports/tree/testing/telegram-desktop/APKBUILD
+# - https://github.com/void-linux/void-packages/blob/master/srcpkgs/telegram-desktop/template
+
+let
+  tg_owt = callPackage ./tg_owt.nix {
+    abseil-cpp = abseil-cpp.override {
+      cxxStandard = "17";
+    };
+  };
+  # Aarch64 default gcc9 will cause ICE. For reference #108305
+  env = if stdenv.isAarch64 then gcc10Stdenv else stdenv;
+in
+env.mkDerivation rec {
+  inherit pname version;
+  # Note: Update via pkgs/applications/networking/instant-messengers/telegram/tdesktop/update.py
+
+  # Telegram-Desktop with submodules
+  src = fetchFromGitHub {
+    owner = "telegramdesktop";
+    repo = "tdesktop";
+    rev = "v${version}";
+    fetchSubmodules = true;
+    inherit sha256;
+  };
+
+  postPatch = ''
+    substituteInPlace Telegram/ThirdParty/libtgvoip/os/linux/AudioInputALSA.cpp \
+      --replace '"libasound.so.2"' '"${alsa-lib}/lib/libasound.so.2"'
+    substituteInPlace Telegram/ThirdParty/libtgvoip/os/linux/AudioOutputALSA.cpp \
+      --replace '"libasound.so.2"' '"${alsa-lib}/lib/libasound.so.2"'
+    substituteInPlace Telegram/ThirdParty/libtgvoip/os/linux/AudioPulse.cpp \
+      --replace '"libpulse.so.0"' '"${libpulseaudio}/lib/libpulse.so.0"'
+    substituteInPlace Telegram/lib_webview/webview/platform/linux/webview_linux_webkit_gtk.cpp \
+      --replace '"libwebkit2gtk-4.0.so.37"' '"${webkitgtk}/lib/libwebkit2gtk-4.0.so.37"'
+    substituteInPlace Telegram/SourceFiles/core/application.cpp \
+      --replace "cExeDir() + cExeName()" '"telegram-desktop"'
+    substituteInPlace Telegram/SourceFiles/platform/linux/specific_linux.cpp \
+      --replace "cExeDir() + cExeName()" '"telegram-desktop"'
```

Comment:
Can you please explain in more detail, what exactly do these lines fix?

Reply:
I think these substitutions are supposed to modify the autostart .desktop file.
Currently, the autostart file will contain a store path:
```Exec=/nix/store/jqji1v34cl1rpayfjd9i0c8db8rcpys6-telegram-desktop-3.6.0/bin/.telegram-desktop-wrapped -workdir /home/arne/.local/share/TelegramDesktop/ -autostart```
After substituting, it should contain `Exec=telegram-desktop -workdir ....`